### ISSUE TITLE
ng-repeat with filter and track by $index and "::" not working 

### ISSUE THREAD

**Issue by squadwuschel** [angular/angular.js#15175]
Bug Report for 

ng-repeat with filter and track by $index and "::" not working right.

**What is the current behavior?**

I want to filter a list of Items and I don't want a watcher on every list entry so I use "::" for my list Entries like:

     <label>Search: <input ng-model="searchText"></label>
     <tr ng-repeat="friend in friends | filter:searchText track by $index">
        <td>{{::friend.name}}</td>
        <td>{{::friend.phone}}</td>
      </tr>

that works fine until I start filtering, then the result is not what it should be.

![filter](https://cloud.githubusercontent.com/assets/1572878/18741443/91c89426-80af-11e6-82d7-87216b09eabe.png)

If you use the following plunker created from the AngularJs Website demo and just added the above track by $index and "::" to the first List and when you there search for "Mike" you will see the result shown above in the Screenshot.

https://plnkr.co/edit/xMgazcv4ACl5GhpNNdTn?p=preview

**What is the expected behavior?**

The Filter Should work right even on "::" assigned List Entries.

**Comment by gkalpak** [comment]
This is expected behavior. 

`track by $index` tells Angular that if the index matches, then it is the same items as before. What this means is that the first `<tr>` will always preceived by Angular as corresponding to the same `friend`. Which in turn means that Angular will reuse the same DOM element and the same scope (for performance reasons). Furthermore, you are telling Angular that once it evaluates the `::` bindngs, it should not evaluate them again. Since the scope remains the same and it has no bindings to evaluate, you don't see the changs in the view (even if the underlying data changes).

**tl;dr**

Don't use `track by $index` with `::` bindings if you need to modify the entries in any way.
(Most of the time, tracking by `$index` is not  a good idea anyway; for example animations won't work as expected.)
